[CES-862] colorado implementation

ImaraSpeek opened the pull request:
Add flake8 error codes to every styling item.
Add the resources required to set up CI/CD and checking for coding styles
Add a readme.md in how to use this in the python folder
(basically this is just a reflection of everything I've implemented in our own opinicus colorado/master_imx6

Josnoww commented on resources/python/run_shellcheck.sh:
```diff
@@ -0,0 +1,42 @@
+#!/bin/sh
+#
+# Copyright (C) 2019 Ultimaker B.V.
+#
+# SPDX-License-Identifier: LGPL-3.0+
+
+# This script is mandatory in a repository, to make sure the shell scripts are correct and of good quality.
+
```
I think we need to get rid of this license.
@SanneMarx ?

Josnoww commented on resources/python/.flake8:
```diff
@@ -0,0 +1,39 @@
+# max McCabe complexity allowed
+max-complexity = 10
+
+ignore =
+  # E266 — too many leading '#' for block comment
+  # Ignore as this is the standard style to doc classes in UM
+  E266
+  # E501 - Max line length exceeded
+  # Ignore as we do not want to set a hard limit, just a guideline
+  E501
+  # CFQ003 - flake8-functions function is not pure
+  # unsure what this checks for
+  CFQ003
+  # CFQ004 - flake8-functions return number
+  # we don't have a set limit on the amount of return statements
+  CFQ004
+
+  # TEMPORARILY HIDING VIOLATIONS
+  # N802 - function name 'someCamelCaseFunction' should be lowercase
+  # Will be ignored until we have decided to change style with sline code
+  N802
+  # E731 do not assign a lambda expression, use a def
+  E731
+
+
+exclude =
+    # No need to traverse our git directory
+    .git,
+    # There's no value in checking cache directories
+    __pycache__,
+    # There's no value in checking the build directory
+    _build_armhf
```
Not sure if intended, but the indentation for ignore and exclude is not the same depth

Josnoww reviewed (commented):
Left a few comments. Seems quite good overall

nallath commented on general/generic_code_conventions.md:
```diff
@@ -104,12 +104,12 @@ switch (variable) {
 The following conventions are used for C/C++ naming. 
- * **variables**: `lower_case_with_underscores`
- * **functions**: `lowerCamelCase`
+ * **variables**: `lower_case_with_underscores` (N806, N815 flake8 pep8-naming)
+ * **functions**: `lowerCamelCase` (N802 flake8 pep8-naming)
```
The Cura team currently has that.

Ghostkeeper replied:
(And hopefully all other code too.)

Ghostkeeper commented on resources/python/.editorconfig:
```diff
@@ -0,0 +1,87 @@
+ij_python_spaces_around_additive_operators = true
+ij_python_spaces_around_assignment_operators = true
+ij_python_spaces_around_bitwise_operators = true
+ij_python_spaces_around_eq_in_keyword_argument = false
```
This is contrary to the code style: Spaces around binary operators.

ImaraSpeek replied:
This is specifically about spaces around eq in keyword arguments `funct(parameter=default)`

Ghostkeeper commented on resources/python/.editorconfig:
```diff
@@ -0,0 +1,87 @@
+ij_python_spaces_around_assignment_operators = true
+ij_python_spaces_around_bitwise_operators = true
+ij_python_spaces_around_eq_in_keyword_argument = false
+ij_python_spaces_around_eq_in_named_parameter = false
```
This is contrary to the code style: Spaces around binary operators.

Ghostkeeper commented on resources/python/.gitlab-ci.yml:
```diff
@@ -0,0 +1,93 @@
+# Copyright (C) 2019 Ultimaker B.V.
+
+include:
+  - project: ultimaker/embedded/prime-jedi
```
Maybe specify in the file name that this is specific to embedded projects, since it apparently is?

Ghostkeeper commented on resources/python/build_for_ultimaker.sh:
```diff
@@ -0,0 +1,125 @@
+
+set -eu
+
+ARCH="armhf"
```
Cura runs on x64 architectures. I think this is specific to certain projects and should probably just be committed to those projects rather than this one, or at least marked as such.

Ghostkeeper commented on resources/python/ci/dead_code_analysis.sh:
```diff
@@ -0,0 +1,7 @@
+
+set -eu
+
+vulture --min-confidence 95 "griffin"
```
Error: "griffin" package doesn't exist. Not in the Cura repository anyway.

Was this meant to be specific to certain repositories? Perhaps it's better to mark it as such or order them in folders as such.

Ghostkeeper commented on resources/python/build_for_ultimaker.sh:
```diff
@@ -0,0 +1,125 @@
+#!/bin/bash
+#
+# Copyright (C) 2019 Ultimaker B.V.
+#
```
Shall we make a variant of this that can run on MacOS and Windows, for those developers that have to work on those platforms? It's intended for the developers to be able to run locally, after all.

Ghostkeeper commented on resources/python/ci/pytest.sh:
```diff
@@ -0,0 +1,19 @@
+rm -rf ./.mypy_cache
+rm -rf ./cov_report
+USE_DUMMY_DBUS=true \
+    PYTHONPATH=:../dbus-interface-lib:../libpalantir:../libPalantir:../charon:../libCharon:../libsmeagol:../libSmeagol:../marvin-service/src \
```
These paths don't exist in my development environment. Maybe you'd have to add these as required repositories in the documentation?

Then again, I also don't really want my tests to fail if there is a mistake in libPalantir or something.

Ghostkeeper commented on resources/python/ci/ultilint.sh:
```diff
@@ -0,0 +1,7 @@
+
+set -eu
+
+USE_DUMMY_DBUS=true PYTHONPATH=:../dbus-interface-lib:../charon:../libSmeagol:../libsmeagol:../marvin-service/src:./ python3 scripts/util/lint_machine_definitions_to_python.py
```
These paths don't exist in my development environment. Do they need to be added as requirements in the documentation?

Ghostkeeper commented on resources/python/make_docker.sh:
```diff
@@ -0,0 +1,44 @@
+
+DOCKER_FILE_CHANGES=$(git diff docker_env/Dockerfile)
+
+REQUIREMENT_FILE_CHANGES=$(git diff requirements.txt)
```
CMakeLists is leading in the Cura repositories (since requirements.txt doesn't work for non-Python code).

nallath replied:
Although I do think that we should swap that; We could also just let the cmakelists call the requirements.txt

jellespijker replied:
I agree with @nallath here. We should make the pip installable dependencies leading and use that in our cmakelists.
That makes settings up a project like Cura in a virtual environment much easier. You only have to export the PATH, LD_PATH and PYTHONPATHS for SIP/Python projects like Savitar, Arcus, Uranium

Ghostkeeper commented on resources/python/make_docker.sh:
```diff
@@ -0,0 +1,44 @@
+
+GIT_HASH="$(git rev-parse HEAD)"
+
+DOCKER_IMAGE_NAME="${1:-opinicus_${GIT_HASH}}"
```
Seems like an image name specific to a certain repository. Should this be added to the Opinicus repository instead?

Ghostkeeper reviewed (changes requested):
Seems like a lot of this is made for a specific project, rather than all of Ultimaker.

If it is for all of Ultimaker, it needs to be made more generic, e.g. to work on Windows and MacOS (using cross-platform tooling like Python, rather than shell scripts, for instance). If not, I suggest we move this code to the specific project it applies to or else into a folder that makes it clear which part of Ultimaker it applies to.

ImaraSpeek commented:
Hey @Ghostkeeper, @nallath and @jellespijker , thank you so much for all your input! I got really excited with all the improvements I made for Opinicus so I wanted to put them somewhere where everyone could use them. For the make_docker script I added a regeneration of the container also when the requirements file was changed, which didn't happen before. I also added the reference.sh script to run the style checks also on local and staged changes as opposed to only committed changes.  We have added some actual flake8 configurations that reflect our coding style using the proper error ignore codes and all and we have added configuration to reflect this styling in the inspection tool for Pycharm. I hope that you can draw some inspiration out of it for now!

Because of some unfortunate branch naming and not knowing this repo was public, we have decided to withdraw the branches for now and use next sprint to summarize the work we have done on our team and create a new branch. Perhaps it is good to have a call with one of you to discuss which parts would be actually helpful for other teams such as Cura, because I may have made it a little to embedded based.
